### PR TITLE
19212 panic after debug sethead different account

### LINKED ISSUE #19212: Panic after debug.setHead() when sending from a different account
Edit: fix submitted in PR #19216.

Panic occurs here:
https://github.com/ethereum/go-ethereum/blob/81ed70015704737b3e5329314a62a3a5aaa74e8a/core/tx_pool.go#L395

After doing the following:

- `eth.sendTransaction({from: A, to: B, value: any})`
- `debug.setHead(back one block)`
- `eth.sendTransaction({from: not A, to: any, value: any})`

What I think is happening:

- After `setHead()`, in tx pool loop `reset()` gets called
- `oldHead` and `newHead` are valid, but one or both of them are no longer in the tx pool
- Therefore one of `add` or `rem` is nil, and prior to [line 395](https://github.com/ethereum/go-ethereum/blob/81ed70015704737b3e5329314a62a3a5aaa74e8a/core/tx_pool.go#L395) no check for nil is done
- Panic on line 395 dereferencing nil

Environment:
```
Geth
Version: 1.8.22-stable
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.11.5
Operating System: darwin
```

Steps to reproduce:

```
# 
# Setup a new geth dev mode environment
#  
$ geth --dev --datadir set_head_bug console
INFO [03-05|21:57:51.684] Maximum peer count                       ETH=25 LES=0 total=25
INFO [03-05|21:57:53.821] Using developer account                  address=0x80f3Cf82e62b7C3948DA0AffDDd2Ed31c10655B5
INFO [03-05|21:57:53.822] Starting peer-to-peer node               instance=Geth/v1.8.22-stable/darwin-amd64/go1.11.5
<snip>
INFO [03-05|21:57:53.829] Writing custom genesis block
INFO [03-05|21:57:53.830] Persisted trie from memory database      nodes=11 size=1.71kB time=102.377µs gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [03-05|21:57:53.830] Initialised chain configuration          config="{ChainID: 1337 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0  ConstantinopleFix: 0 Engine: clique}"
INFO [03-05|21:57:53.830] Initialising Ethereum protocol           versions="[63 62]" network=1337
INFO [03-05|21:57:53.849] Loaded most recent local header          number=0 hash=3a565f…660acd td=1 age=49y10mo2w
<snip>
INFO [03-05|21:57:53.857] Sealing paused, waiting for transactions
Welcome to the Geth JavaScript console!

instance: Geth/v1.8.22-stable/darwin-amd64/go1.11.5
coinbase: 0x80f3cf82e62b7c3948da0affddd2ed31c10655b5
at block: 0 (Thu, 01 Jan 1970 09:00:00 JST)
 datadir: /tmp/set_head_bug
 modules: admin:1.0 clique:1.0 debug:1.0 eth:1.0 miner:1.0 net:1.0 personal:1.0 rpc:1.0 shh:1.0 txpool:1.0 web3:1.0

> eth.blockNumber
0

# 
# Setup a second account
# 
> personal.newAccount()
Passphrase:
Repeat passphrase:
"0x6941474503af6d776c0ca7f881ce399b416ed79c"
> personal.unlockAccount(eth.accounts[1], "", 86400)
true

# 
# Send the new account Ether twice to get us to block number 2
# 
> eth.sendTransaction({from: eth.accounts[0], to: eth.accounts[1], value: 1000000})
INFO [03-05|21:59:07.338] Setting new local account                address=0x80f3Cf82e62b7C3948DA0AffDDd2Ed31c10655B5
INFO [03-05|21:59:07.338] Submitted transaction                    fullhash=0x41979183f48b61b953c585be7fa6f3da9214707dc0bf904813e9580c27cc9c86 recipient=0x6941474503af6D776C0ca7f881Ce399B416Ed79C
INFO [03-05|21:59:07.338] Commit new mining work                   number=1 sealhash=e028a8…e70d2e uncles=0 txs=0 gas=0 fees=0 elapsed=94.44µs
"0x41979183f48b61b953c585be7fa6f3da9214707dc0bf904813e9580c27cc9c86"
INFO [03-05|21:59:07.339] Sealing paused, waiting for transactions
> INFO [03-05|21:59:07.339] Commit new mining work                   number=1 sealhash=4ce0e3…52aa35 uncles=0 txs=1 gas=21000 fees=2.1e-14 elapsed=382.625µs
INFO [03-05|21:59:07.339] Successfully sealed new block            number=1 sealhash=4ce0e3…52aa35 hash=c7d1b5…885c81 elapsed=742.391µs
INFO [03-05|21:59:07.340] 🔨 mined potential block                  number=1 hash=c7d1b5…885c81
INFO [03-05|21:59:07.340] Commit new mining work                   number=2 sealhash=e716ed…aca8f1 uncles=0 txs=0 gas=0     fees=0       elapsed=306.807µs
INFO [03-05|21:59:07.340] Sealing paused, waiting for transactions

> eth.blockNumber
1
> eth.sendTransaction({from: eth.accounts[0], to: eth.accounts[1], value: 1000000})
INFO [03-05|21:59:13.082] Submitted transaction                    fullhash=0xd6b4a177a7e581156660628103847d9883c78ec3bdb00e4e955107125609d2b7 recipient=0x6941474503af6D776C0ca7f881Ce399B416Ed79C
"0xd6b4a177a7e581156660628103847d9883c78ec3bdb00e4e955107125609d2b7"
> INFO [03-05|21:59:13.082] Sealing paused, waiting for transactions
INFO [03-05|21:59:13.082] Commit new mining work                   number=2 sealhash=fd477b…0d36c2 uncles=0 txs=0 gas=0     fees=0       elapsed=174.367µs
INFO [03-05|21:59:13.082] Commit new mining work                   number=2 sealhash=53d144…70cc29 uncles=0 txs=1 gas=21000 fees=2.1e-14 elapsed=460.98µs
INFO [03-05|21:59:13.083] Successfully sealed new block            number=2 sealhash=53d144…70cc29 hash=ee9dfb…c108b5 elapsed=733.969µs
INFO [03-05|21:59:13.083] 🔨 mined potential block                  number=2 hash=ee9dfb…c108b5
INFO [03-05|21:59:13.083] Commit new mining work                   number=3 sealhash=41090d…21df11 uncles=0 txs=0 gas=0     fees=0       elapsed=503.486µs
INFO [03-05|21:59:13.083] Sealing paused, waiting for transactions
>
> eth.blockNumber
2
>
>

# 
# Roll back to block number 1
# Because we sent Ether twice to the second account, it still
# has Ether for the next step 
# 
> debug.setHead("0x1")
WARN [03-05|21:59:31.606] Rewinding blockchain                     target=1
INFO [03-05|21:59:31.607] Loaded most recent local header          number=1 hash=c7d1b5…885c81 td=3 age=24s
INFO [03-05|21:59:31.607] Loaded most recent local full block      number=1 hash=c7d1b5…885c81 td=3 age=24s
INFO [03-05|21:59:31.607] Loaded most recent local fast block      number=1 hash=c7d1b5…885c81 td=3 age=24s
null
> eth.blockNumber
1

# 
# Send any transaction that's not from the first account
# 
> eth.sendTransaction({from: eth.accounts[1], to: eth.accounts[0], value: 1})
INFO [03-05|21:59:49.138] Setting new local account                address=0x6941474503af6D776C0ca7f881Ce399B416Ed79C
INFO [03-05|21:59:49.138] Submitted transaction                    fullhash=0x78d7a5672445ac66cc34177b931cc95bc59b0ceb9fac543500d470b9a4e08062 recipient=0x80f3Cf82e62b7C3948DA0AffDDd2Ed31c10655B5
INFO [03-05|21:59:49.139] Commit new mining work                   number=2 sealhash=b10065…efac29 uncles=0 txs=0 gas=0     fees=0       elapsed=105.499µs
"0x78d7a5672445ac66cc34177b931cc95bc59b0ceb9fac543500d470b9a4e08062"
INFO [03-05|21:59:49.139] Sealing paused, waiting for transactions
> INFO [03-05|21:59:49.139] Commit new mining work                   number=2 sealhash=5bc527…8ce71a uncles=0 txs=1 gas=21000 fees=2.1e-14 elapsed=329.998µs
INFO [03-05|21:59:49.139] Successfully sealed new block            number=2 sealhash=5bc527…8ce71a hash=d1138a…e77550 elapsed=577.576µs
INFO [03-05|21:59:49.139] 🔨 mined potential block                  number=2 hash=d1138a…e77550
INFO [03-05|21:59:49.140] Commit new mining work                   number=3 sealhash=b1a46f…9d3929 uncles=0 txs=0 gas=0     fees=0       elapsed=368.017µs
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x44b6011]

goroutine 15 [running]:
github.com/ethereum/go-ethereum/core/types.(*Block).NumberU64(...)
	/private/tmp/ethereum-20190204-77575-oobokl/build/_workspace/src/github.com/ethereum/go-ethereum/core/types/block.go:291
github.com/ethereum/go-ethereum/core.(*TxPool).reset(0xc00022da40, 0xc021f07b00, 0xc021f07d40)
	/private/tmp/ethereum-20190204-77575-oobokl/build/_workspace/src/github.com/ethereum/go-ethereum/core/tx_pool.go:395 +0x6e1
github.com/ethereum/go-ethereum/core.(*TxPool).loop(0xc00022da40)
	/private/tmp/ethereum-20190204-77575-oobokl/build/_workspace/src/github.com/ethereum/go-ethereum/core/tx_pool.go:314 +0x43c
created by github.com/ethereum/go-ethereum/core.NewTxPool
	/private/tmp/ethereum-20190204-77575-oobokl/build/_workspace/src/github.com/ethereum/go-ethereum/core/tx_pool.go:278 +0x677
```

### PR DESCRIPTION
**Pull request by shoenseiwaso**
Resolves #19212.

Fixes the panic by checking for nil `rem` and `add`, and adding the non-nil block's transactions into `discarded` or `included` as appropriate. **Note: no tests yet in `tx_pool_test.go` because I wasn't sure where to start. Some guidance would be appreciated.**

Steps to verify the fix:

```
$ make geth
$ mkdir set_head_bug
$ build/bin/geth --dev --datadir set_head_bug console
> eth.blockNumber;
> personal.newAccount();
Passphrase:
Repeat passphrase:
> personal.unlockAccount(eth.accounts[1], "", 86400);
> eth.sendTransaction({from: eth.accounts[0], to: eth.accounts[1], value: 1000000});
> eth.sendTransaction({from: eth.accounts[0], to: eth.accounts[1], value: 1000000});
> debug.setHead("0x1");
> eth.sendTransaction({from: eth.accounts[1], to: eth.accounts[0], value: 1});
```

No panic after this last command, which is currently the case on `master` as reported in #19212.

### CODE REVIEW

File: core/tx_pool.go [hunk 1]
```diff
@@ -392,6 +392,18 @@ func (pool *TxPool) reset(oldHead, newHead *types.Header) {
...
+			// handle cases where rem or add could come up nil
+			if rem == nil && add == nil {
+				return
+			} else if rem != nil && add == nil {
+				discarded = append(discarded, rem.Transactions()...)
+				return
+			} else if rem == nil && add != nil {
+				included = append(included, add.Transactions()...)
+				return
+			}
```
Comment (holiman) [inline core/tx_pool.go]: The change might be a bit cleaner written like this:
```golang
	// handle cases where rem or add could come up nil
	if rem == nil || add == nil {
		if rem != nil{
			discarded = append(discarded, rem.Transactions()...)
		}
		if add != nil{
			included = append(included, add.Transactions()...)
		}
		return
	}
```

**Comment by shoenseiwaso** [comment]
Good suggestion, I've incorporated it.

**Comment by shoenseiwaso** [comment]
Fixed the linting issues, ready for another look.

Do we want to add a regression test for this issue? If so, could you point me in the direction of where to add it?

**Comment by adamschmideg** [comment]
We're taking another approach and leaving this PR open as a reminder

**Comment by holiman** [comment]
Superseded by https://github.com/ethereum/go-ethereum/pull/19308

**Pull request closed without merging**